buildSrc/src/main/java/org/elasticsearch/gradle/release/GenerateReleaseNotesTask.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.gradle.release;
+
+import org.elasticsearch.gradle.Version;
+import org.elasticsearch.gradle.VersionProperties;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.file.ConfigurableFileCollection;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.file.RegularFileProperty;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+import org.gradle.api.tasks.InputFiles;
+import org.gradle.api.tasks.OutputFile;
+import org.gradle.api.tasks.TaskAction;
+import org.gradle.api.tasks.util.PatternSet;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+public class GenerateReleaseNotesTask extends DefaultTask {
+    private static final Logger LOGGER = Logging.getLogger(GenerateReleaseNotesTask.class);
+
+    private final ConfigurableFileCollection changelogs = getProject().getObjects().fileCollection();
+    private final RegularFileProperty releaseNotesFile = getProject().getObjects().fileProperty();
+    private final RegularFileProperty releaseHighlightsFile = getProject().getObjects().fileProperty();
+    private final RegularFileProperty breakingChangesFile = getProject().getObjects().fileProperty();
+
+    public GenerateReleaseNotesTask() {
+        final Version version = VersionProperties.getElasticsearchVersion();
+
+        this.changelogs.setFrom(
+            getProject().getLayout()
+                .getProjectDirectory()
+                .dir("docs/changelog")
+                .getAsFileTree()
+                .matching(new PatternSet().include("**/*.yml", "**/*.yaml"))
+                .getFiles()
+        );
+
+        this.releaseNotesFile.set(
+            getProject().getLayout()
+                .getProjectDirectory()
+                .file(String.format("docs/reference/release-notes/%d.%d.asciidoc", version.getMajor(), version.getMinor()))
+        );
+
+        this.releaseHighlightsFile.set(
+            getProject().getLayout().getProjectDirectory().file("docs/reference/release-notes/highlights.asciidoc")
+        );
+
+        this.breakingChangesFile.set(
+            getProject().getLayout()
```

Comment:
In case you need `ProjectLayout` in a task please inject it via constructor like described above with `ObjectFactory`